exist-core/src/main/java/org/exist/xquery/functions/map/MapType.java

```diff
@@ -1,169 +1,252 @@
+    public MapType(final XQueryContext context, @Nullable final Collator collator, final AtomicValue key, final Sequence value) {
         super(context);
-        if (collation == null)
-            {this.map = PersistentHashMap.EMPTY;}
-        else
-            {this.map = PersistentTreeMap.create(getComparator(collation), null);}
+        this.map = newMap(collator).put(key, value);
         this.type = key.getType();
-        this.map = this.map.assoc(key, value);
     }
 
-    protected MapType(XQueryContext context, IPersistentMap<AtomicValue, Sequence> other, int type) {
+    public MapType(final XQueryContext context, @Nullable final Collator collator, final Iterable<Tuple2<AtomicValue, Sequence>> keyValues) {
+        this(context, collator, keyValues.iterator());
+    }
+
+    public MapType(final XQueryContext context, @Nullable final Collator collator, final Iterator<Tuple2<AtomicValue, Sequence>> keyValues) {
+        super(context);
+
+        // bulk put
+        final IMap<AtomicValue, Sequence> map = newMap(collator).linear();
+        keyValues.forEachRemaining(kv -> map.put(kv._1, kv._2));
+        this.map = map.forked();
+
+        //TODO(AR) this is incorrect all types to be checked!
+        final long size = this.map.size();
+        if (size > 0) {
+            setKeyType(this.map.nth(size - 1).key().getType());
+        }
+    }
+
+    public MapType(final XQueryContext context, final IMap<AtomicValue, Sequence> other, @Nullable final Integer type) {
         super(context);
+
+        if (other.isLinear()) {
+            throw new IllegalArgumentException("Map must be immutable, but linear Map was provided");
+        }
+
         this.map = other;
-        this.type = type;
-    }
-
-    public void add(AbstractMapType other) {
-        if (other.size() == 1) {
-            setKeyType(other.getKey().getType());
-            map = map.assoc(other.getKey(), other.getValue());
-        } else if (other.size() > 0) {
-            setKeyType(other.getKeyType());
-            if (map instanceof PersistentHashMap) {
-                ITransientMap<AtomicValue, Sequence> tmap = ((PersistentHashMap)map).asTransient();
-                for (final Map.Entry<AtomicValue, Sequence> entry : other) {
-                    tmap = tmap.assoc(entry.getKey(), entry.getValue());
+
+        if (type != null) {
+            this.type = type;
+        } else {
+            //TODO(AR) this is incorrect all types to be checked!
+            final long size = this.map.size();
+            if (size > 0) {
+                setKeyType(this.map.nth(size - 1).key().getType());
+            }
+        }
+    }
+
+    public void add(final AbstractMapType other) {
+        setKeyType(other.key() != null ? other.key().getType() : Type.ANY_TYPE);
+
+        if(other instanceof MapType) {
+            //TODO(AR) is the union in the correct direction i.e. keys from `other` overwrite `this`
```

Comment:
Are you planning to remove / tackle these TODOs before this PR is merged?
We should have an answer especially for this one _before_ merging.

Reply:
I have resolved this now.